Support MaterializedView in Cassandra connector

ebyhr opened the pull request:
CachingCassandraSchemaProvider's `loadTable` method may not be efficient since it gets all tables in the schema.

#7363

ebyhr commented:
Rebased. product-test on upstream seems fail with following message. 
```
...
presto-worker_1       | 2017-06-18T13:42:20.466Z	INFO	Code-Cache-GC-Trigger	com.facebook.presto.server.CodeCacheGcTrigger	Triggering GC to avoid Code Cache eviction bugs
presto-master_1       | 2017-06-18T13:42:24.122Z	INFO	Code-Cache-GC-Trigger	com.facebook.presto.server.CodeCacheGcTrigger	Triggering GC to avoid Code Cache eviction bugs
The job exceeded the maximum time limit for jobs, and has been terminated.
```

cawallin commented on presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java:
```diff
@@ -232,4 +239,42 @@ public void testNationJoinRegion()
+                                "{\"a\":1,\"b\":2}", "[4,5,6]", "this is a text value", Timestamp.valueOf("9999-12-31 23:59:59"),
+                                "d2177dd0-eaa2-11de-a572-001b779c76e3", "01234567-0123-0123-0123-0123456789ab",
+                                "abc", String.valueOf(Long.MAX_VALUE)));
+    }
```
Please drop the materialized view at the end of the test

cawallin replied:
Can you also add tests for
* Selecting by partition
* more complicated select
* INSERT, if it's supported -- if not, we probably want to explicitly disallow insert in Presto with materialized views

ebyhr commented:
Rebased but the travis still fails.

cawallin commented on presto-cassandra/src/main/java/com/facebook/presto/cassandra/NativeCassandraSession.java:
```diff
@@ -171,6 +173,9 @@ public String getCaseSensitiveSchemaName(String caseInsensitiveSchemaName)
         for (TableMetadata table : keyspace.getTables()) {
             builder.add(table.getName());
         }
+        for (MaterializedViewMetadata materializedView : keyspace.getMaterializedViews()) {
```
Also, "Currently, only simple SELECT statements are supported, but a ticket has been filed to add support for more complex SELECT statements
WHERE clauses, ORDER BY, and functions aren't available with materialized views" (https://www.datastax.com/dev/blog/new-in-cassandra-3-0-materialized-views) -- That suggests that Presto doesn't support the initial, Cassandra 3.0 version of materialized views on partitioning keys, since we push down a where clause in that case. Can you only add materialized views here if the Cassandra version is >= whatever Cassandra version adds full materialized view support? You can get the version via `getCassandraVersion()` in `CassandraSession`

ebyhr replied:
keyspace.getMaterializedViews() on pre-3.0 returns empty object without error like this.
<img width="858" alt="screen shot 2017-08-11 at 14 45 27" src="https://user-images.githubusercontent.com/6237050/29219110-0237ee3c-7ef2-11e7-93ea-71eddcaa410e.png">

ebyhr replied:
I'm testing on 3.0.0 but all queries finish successfully. Would you give me the query which should be failed in presto? The JIRA page says that Fix Version is 3.0.0 rc1, so the negative versions might only be 3.0.0 alpha1, 3.0.0 beta1 and 3.0.0 beta2.

cawallin replied:
OK great, it's ok to not support the beta versions. So we don't need version checking here.

cawallin commented on presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java:
```diff
@@ -232,4 +239,85 @@ public void testNationJoinRegion()
+    public void testSelectAllTypeParitioningMateliarizedView()
+            throws SQLException
+    {
+        String mvName = format("%s_partitioned_mv", CASSANDRA_ALL_TYPES.getName());
```
materializedViewName -- the Presto coding style prefers not using abbreviations

ebyhr commented:
Materialized View on Cassandra sometimes returns wrong value. My code already checks system.built_views.status_replicated, though. I'm investigating the reason. 
```
java.lang.AssertionError: Could not find rows:
[, -9223372036854775808, [B@61d9dddf, false, 0.0, 4.9E-324, 1.4E-45, [0], 0.0.0.0, -2147483648, [0], {"\u0000":-2147483648,"a":0}, [0], , 1970-01-01 05:30:00.0, d2177dd0-eaa2-11de-a572-001b779c76e3, 01234567-0123-0123-0123-0123456789ab, , -9223372036854775808]

actual rows:
[, -9223372036854775808, [B@16902f7b, false, 0.0, 4.9E-324, 1.4E-45, [0], 0.0.0.0, -2147483648, [0], {"\u0000":-2147483648,"a":0}, [0], , 1970-01-01 00:00:00.0, d2177dd0-eaa2-11de-a572-001b779c76e3, 01234567-0123-0123-0123-0123456789ab, , -9223372036854775808]
```

gregdingle commented:
What's the status of this PR? We'd love to be able to use it if it's working, or we may be able to fix it if not.

ebyhr commented:
@gregdingle This PR was under review status, but unfortunately cawallin had left full time contributing.

@kokosing @findepi Could you review this PR?

kokosing commented on presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSession.java:
```diff
@@ -54,6 +54,8 @@ CassandraTable getTable(SchemaTableName schemaTableName)
 
     List<CassandraPartition> getPartitions(CassandraTable table, List<Object> filterPrefix);
 
+    boolean isMaterializedView(String caseInsensitiveSchemaName, String caseInsensitiveTableName);
```
why not to use `SchemaTableName`? So far every name in Presto is case insensitive.

kokosing commented on presto-cassandra/src/main/java/com/facebook/presto/cassandra/NativeCassandraSession.java:
```diff
@@ -266,13 +271,15 @@ private KeyspaceMetadata getKeyspaceByCaseInsensitiveName(String caseInsensitive
 
-    private static TableMetadata getTableMetadata(KeyspaceMetadata keyspace, String caseInsensitiveTableName)
+    private static AbstractTableMetadata getTableMetadata(KeyspaceMetadata keyspace, String caseInsensitiveTableName)
     {
```
What would you say to rewrite this to streams?

```
    private static AbstractTableMetadata getTableMetadata(KeyspaceMetadata keyspace, String caseInsensitiveTableName)
    {
        List<AbstractTableMetadata> tables = Stream.concat(
                keyspace.getTables().stream(),
                keyspace.getMaterializedViews().stream())
                .filter(table -> table.getName().equalsIgnoreCase(caseInsensitiveTableName))
                .collect(toImmutableList());
        if (tables.size() == 0) {
            throw new TableNotFoundException(new SchemaTableName(keyspace.getName(), caseInsensitiveTableName));
        }
        else if (tables.size() == 1) {
            return tables.get(0);
        }
        String tableNames = tables.stream()
                .map(AbstractTableMetadata::getName)
                .sorted()
                .collect(joining(","));
        throw new PrestoException(
                NOT_SUPPORTED,
                format("More than one table has been found for the case insensitive table name: %s -> (%s)",
                        caseInsensitiveTableName, tableNames));
    }
```

ebyhr commented:
Thanks for your review! I'll check the cause of failed product tests.

ebyhr commented:
Updated. I moved `CassandraQueryExecutor` from `getRequirements` to new method since past logic may cause NPE. 

The error message in product test is `The job exceeded the maximum time limit for jobs, and has been terminated` and it seems this commit isn't root cause. Other commits are also terminated by timeout.
[https://travis-ci.org/prestodb/presto/jobs/366033522](https://api.travis-ci.org/v3/job/366033522/log.txt)
[https://travis-ci.org/prestodb/presto/jobs/366384114](https://api.travis-ci.org/v3/job/366384114/log.txt)

gregdingle commented:
@kokosing @ebyhr Thank you for getting involved. What's needed to finish this PR? It looks like the travis check that failed is not reliable in general.

kokosing commented:
@ebyhr can you please rebase, mentioned product tests failures were recently fixed, now your PR should pass the automation.

@electrum Would you like to take a look? I would like to merge this after the release goes out.

kokosing commented:
@ebyhr This is not good ; / It looks that there is a bug.

You need to run product tests manually and see why it does not exit. Here you have instruction how you can set it up: https://github.com/prestodb/presto/blob/master/presto-product-tests/README.md. Also `.travis.yml` is a good place where everything is explained how tests are executed ;)

I remember we used to had such issue with cassandra, and it was related with not closed connection which had some no daemon threads running. Also I saw many stacktraces in the build output, are they expected? Maybe some query failed with exception and connection was not closed for such queries.

ebyhr commented:
@kokosing Thanks for teaching me! Fixed. Though I was testing on local machine, unit tests on IDEA didn't cause this behavior. run_on_docker.sh could reproduce it.

As you wrote, the failed reason was related with not closed connection. I think those stacktraces are expected and shown to assert failed messages.
